vector: add support for vector type

QuerthDP opened the pull request:
This pull request is an implementation of vector data type similar to one used by Apache Cassandra.

The patch contains:
- implementation of vector_type_impl class
- necessary functionalities similar to other data types
- support for serialization and deserialization of vectors
- support for Lua and JSON format
- valid CQL syntax for `vector<>` type
- `type_parser` support for vectors
- expression adjustments such as:
    - add `collection_constructor::style_type::vector`
    - rename `collection_constructor::style_type::list` to `collection_constructor::style_type::list_or_vector`
- vector type encoding (for drivers)
- unit tests
- cassandra compatibility tests
- necessary documentation

Co-authored-by: @janpiotrlakomy 

Fixes https://github.com/scylladb/scylladb/issues/19455

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22488

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/22488)

**Changed Files:**
- [architecture/sstable/sstable3/sstables-3-statistics](http://previews.docs.scylladb.com/scylladb/22488/architecture/sstable/sstable3/sstables-3-statistics)
- [cql/appendices](http://previews.docs.scylladb.com/scylladb/22488/cql/appendices)
- [cql/definitions](http://previews.docs.scylladb.com/scylladb/22488/cql/definitions)
- [cql/json](http://previews.docs.scylladb.com/scylladb/22488/cql/json)
- [cql/non-reserved-keywords](http://previews.docs.scylladb.com/scylladb/22488/cql/non-reserved-keywords)
- [cql/types](http://previews.docs.scylladb.com/scylladb/22488/cql/types)
- [reference/limits](http://previews.docs.scylladb.com/scylladb/22488/reference/limits)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

avikivity commented on cql3/type_json.cc:
```diff
@@ -230,6 +231,32 @@ static bytes from_json_object_aux(const tuple_type_impl& t, const rjson::value&
+    }
+    return vector_type_impl::build_value(std::move(raw_vector), t.value_length_if_fixed());
+}
+
```
Is this tested somewhere?

QuerthDP replied:
This functionality is tested [here](https://github.com/scylladb/scylladb/blob/bd6d4a004c520132af4074c3eaf56f01207fae42/test/boost/json_cql_query_test.cc#L721-L746). 
I believe this test is sufficient, but we might adjust/expand it, if there are still any concerns.

nyh replied:
Personally, I don't understand why write any functionality tests in a Boost test. It should be in cqlpy (I'd create a new file test/cqlpy/test_type_vector.py for it), so you'll be able to compare the functionality to that in Cassandra.
Although specifically for JSON output and input, we actually have a lot of these tests in test/cqlpy/test_json.py. For example see test_tojson_timestamp which checked tojson() of the "timestamp" type - which we used to have a bug in.

QuerthDP replied:
This boost test have been added due to the reasoning that other data types have this kind of test here as well.

We are now aware that this PR lacks the reasonable amount of cqlpy tests, and it is our high priority TODO to deliver them in a close future. Will keep you updated!

avikivity commented on lang/lua.cc:
```diff
@@ -735,6 +735,43 @@ struct from_lua_visitor {
+        }
+        return make_vector_value(t.shared_from_this(), std::move(elements));
+    }
+
```
And this? Also need to documented how a vector value is translated to lua.

QuerthDP replied:
This one is tested [here](https://github.com/scylladb/scylladb/blob/cc3cac241ef3757bbc2ba74bd8b464baed213cce/test/boost/user_function_test.cc#L722-L745).
We have adjusted the [docs](https://github.com/scylladb/scylladb/blob/bd6d4a004c520132af4074c3eaf56f01207fae42/docs/dev/lua-type-mapping.md?plain=1#L131-L133), but I've just noticed it might need some more explanation (similar to the tuple type, as we also check the number of elements).

avikivity commented on types/types.cc:
```diff
@@ -1610,6 +1632,124 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+            return cmp;
+        }
+    }
+
```
Please verify that this ordering is compatible with Cassandra.

nyh replied:
This is why I encourage people to write all tests that can be exercised through CQL in test/cqlpy and not in the test/boost cql_env framework. The latter always leaving you wondering whether the behavior you just wrote a test for actually matches Cassandra's behavior. The former allows you to really be sure that Scylla's behavior is identical to Cassandra and not incompatible with Cassandra for no good read.

I saw you translated Cassandra's unit tests and put it in test/cqlpy/cassandra_tests, which is great, but you can also add new cqlpy tests, just don't put them in cassandra_tests (a new file test/cqlpy/test_type_vector.py would be a good place for new tests for the vector type).

avikivity commented on types/types.cc:
```diff
@@ -1610,6 +1632,124 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+    }
+
+    return data_value::make(t.shared_from_this(), std::make_unique<vector_type_impl::native_type>(std::move(ret)));
+}
```
Also: check the serialization protocol. We should be able to exchange sstables with Cassandra.

janpiotrlakomy replied:
The serialization protocol is correct and it is copied from the one found in Cassandra's code. It is, however, different from the one found in the [native_protocol_v5.spec](https://github.com/apache/cassandra/blob/15ed18e9d49f48e88f40b90c156248b8b697c7e2/doc/native_protocol_v5.spec#L1210-L1215).

When serializing a vector with variable-length  elements, each element is preceded by its length. The difference between  [native_protocol_v5.spec](https://github.com/apache/cassandra/blob/15ed18e9d49f48e88f40b90c156248b8b697c7e2/doc/native_protocol_v5.spec#L1210-L1215) and Cassandra is that the protocol specification asks to serialize the length as a 4 byte integer, but Cassandra serializes it as a unsigned varint.

We think its just an error in the documentation.

avikivity replied:
See my earlier comment about Cassandra's cqlsh being unable to deserialize vectors of variable-size types. It's better to forbid them for now.

avikivity commented on types/types.cc:
```diff
@@ -2852,6 +3004,11 @@ struct from_string_visitor {
+    bytes operator()(const vector_type_impl& t) {
+        // FIXME:
+        abort();
+        return bytes();
```
?

QuerthDP replied:
Although we are aware that this implementation (or rather lack of it) is not fully compatible with Apache Cassandra,
we decided to keep it that way as the functionality is truly niche, and should not be introduced in advance, while collections does not support it yet.

It has been adjusted to match Scylla's behaviour of trying to insert collections from string, which leads to throwing an exception (`Invalid STRING constant ...`) caused by untyped constant not being able to be assigned to neither vectors nor collections. As this exception is thrown before we would use the `from_string_visitor`, the `abort()` function is never called here.

QuerthDP replied:
We believe that the implementation for both collections and vector type will be delivered in future.

avikivity replied:
Ah, I see it's copied from the collection branch below. Which has a similarly inexpressive comment.

Can it be triggered? What code path leads here?

QuerthDP replied:
I believe this code is never triggered. 

The only possible way I have cought when I could get the `abort()` called was through the `untyped_constant_prepare_expression` function.
But it has been fixed by setting `untyped_constant_test_assignment` result to `NOT_ASSIGNABLE` for vector type, which now leads to throwing an `invalid_request_exception`, so that the `from_string` function is not reached in this case.

avikivity replied:
Ok. It's not good to rely on such non-local dependencies though.

Better to replace it with an exception. Since it's a preexisting problem, we can do it after merging (and apply to other aborts in this function).

avikivity commented on cql3/Cql.g:
```diff
@@ -2035,6 +2036,11 @@ tuple_type [bool internal] returns [shared_ptr<cql3::cql3_type::raw> t]
 
+vector_type [bool internal] returns [shared_ptr<cql3::cql3_type::raw> pt]
+    : K_VECTOR '<' t=comparator_type[internal] ',' d=INTEGER '>'
+        { if (t) { $pt = cql3::cql3_type::raw::vector(t, std::stoul($d.text)); } }
```
What if !t?

I think you can just drop the condition, I don't see how it can happen.

avikivity replied:
And what's the internal parameter?

QuerthDP replied:
The meaning of internal paremeter is explained [here](https://github.com/scylladb/scylladb/blob/bd6d4a004c520132af4074c3eaf56f01207fae42/cql3/Cql.g#L1940-L1949).

AFAIU this parameter yet only indicates if the type `empty` can be used internally as the type of elements.
I used it there in an early stage of vector type implementation, as other composed data types used it as well.
Now I think it might be unnecessary here, and we could simply set it to `false` as vectors should not support empty type inside (but I might be mistaken here).

Please let me know what do you think. (@janpiotrlakomy @piodul could you help with that?)

QuerthDP replied:
Thanks for explanation!
I will change it accordingly not to support empty.

avikivity commented on cql3/cql3_type.cc:
```diff
@@ -305,6 +305,56 @@ class cql3_type::raw_tuple : public raw {
+    }
+
+    virtual bool supports_freezing() const override {
+        return true;
```
Freezing is meaningful only to collections whose elements can be stored as independent cells with their own timestamps and tombstones, which isn't the case here.

QuerthDP replied:
I can agree, but we have decided to stick with [Cassandra's implementation](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/CQL3Type.java#L911-L915), where this exact same method returns `true` as well.

Nevertheless, it does not provide any impact on vector itself.
With or without declaring the vector `frozen<vector<...>>` it is displayed (for example in `DISCRIBE [table]` result) without the `frozen<...>` tag.
Furthermore, the types of elements inside vector, are not being [implicitly frozen](https://github.com/zpp-2024-vector-search/scylladb/pull/5#discussion_r1863365070) as well.

Moreover, there is a [method in Cassandra](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/CQL3Type.java#L634-L637) which (AFAIU) states that the vector is **always** frozen (but why wouldn't it behave particularly like one?).

I believe we have to decide on one case or another, how we would like our implementation to behave.
Yet we have sticked to the implementation Cassandra introduced, to match their behaviour as a starting point.

cc: @piodul

avikivity replied:
Ok. Please make sure the tests cover frozen/non-frozen declarations in CQL.

avikivity replied:
We do know it has no effect. The vector type is always frozen.

nyh replied:
I still don't understand what this function is supposed to do - is it to allow the "frozen<vector<...>>" syntax? Why do we want to allow such a silly syntax? Or is this function used for something else?

QuerthDP replied:
To sum up:
- we need to implement the `supports_freezing` function as it is a virtual function and needs to be implemented somehow
- it is called to check if we could call `freeze()` on vector (e.g. when we want to store `tuple<vector<...>, ...>` or `frozen<list<vector<...>>>`
- we should be able to call `freeze()` (which does not have any effect anyway as vector is always frozen)
- as vector is always frozen, then why would the `supports_freezing` return `false`?
- it is identically implemented in Cassandra

Furthermore, I don't think it has to do anything with allowance of `frozen<vector<...>>` syntax.
It just does not provide any effect (as `frozen<>` just calls the `freeze()` function).

If we wanted to disallow the `frozen<vector<...>>` syntax, we would have not to override the `freeze()` function, and change it's exception message in implementation of `cql3_type::raw::freeze()`, **BUT** it wouldn't match Cassandra's implementation then.

avikivity commented on cql3/cql3_type.cc:
```diff
@@ -305,6 +305,56 @@ class cql3_type::raw_tuple : public raw {
+
+    virtual cql3_type prepare_internal(const sstring& keyspace, const data_dictionary::user_types_metadata& user_types) override {
+        if (!is_frozen()) {
+            freeze();
```
see? it loses its meaning

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -70,6 +70,7 @@ unresolved_identifier::~unresolved_identifier() = default;
 static cql3::raw_value do_evaluate(const bind_variable&, const evaluation_inputs&);
 static cql3::raw_value do_evaluate(const tuple_constructor&, const evaluation_inputs&);
```
> In particular, this separates the implementation from collections,
which could've been quite misleading and inelegant.

But is there any practical difference from collection_constructor? This adds a lot of noise.

QuerthDP replied:
I do believe it is crucial to add the `vector_constructor`, regardless.

1. First of all, if we used `collection_constructor` the way it is implemented now, we would have to use the `do_evaluate` function with it. 
There emerges a problem, as if we come into `evaluate_list` (as we would be redirected here, while literals for vectors and the lists are the same) then the wrong serialization function would be executed (`serialize_listlike` function differs from the one required for vectors according to [the protocol](https://github.com/apache/cassandra/blob/15ed18e9d49f48e88f40b90c156248b8b697c7e2/doc/native_protocol_v5.spec#L1210-L1215) [1]).
We cannot just simply differ the vector from the list here, because there is no access to the column type from this part of code, it would require major refactoring if we wanted to do it this way.
Moreover, which is a minor issue, the exception messages would not match, as the code would believe we executed the `evaluate_list` for the list, not vector, and `list` is hard coded here in the messages.
    * 💡 We could have introduced new `collection_constructor::style_type` different from the list one and implement crucial functions for vectors in `collection_constructor` the same way as we split maps and sets from lists, **BUT** as mentioned before it is not possible due to the ambiguity of list and vector literals...
2. To resolve the issue, we implemented the distinction in `prepare_expression`, as there is a simple access to the column type through the `receiver`.
3. Furthermore, we needed the split during preparation anyway, because the preparation for vectors and lists slightly differs (mainly due to the fact that lists can be empty), and that we need to produce proper column specification of `vector_type_impl`.
4. On the other hand, I think it could be done the other way around, if we truly don't want to introduce a new constructor, but it would require more changes and does not come naturally to my mind how to do it.

On a final note, I do think clean and tidy code is a good thing to introduce, and splitting the vector from collections is,
I believe an example of it, as it delivers more clarity for future development and maintenance.
Of course, it might not be perfect as we do still learn and gain experience, but I hope it's clear enough what should it do.

[1] the protocol itself has some other issues, which will be mentioned [here](https://github.com/scylladb/scylladb/pull/22488#discussion_r1929087623)

avikivity replied:
I think adding a new style_type would be better (alternatively, switching on collection_constructor::type).

I understand your motivation to have the vector code separate, but my experience with the expression code is that we won't touch collection_constructor again (or at least until someone adds a new form of collections). On the other hand anyone implementing or maintaining a visitor will have to handle yet another top-level case which just leads to repetition. Better to isolate construction in one place.

QuerthDP replied:
Oh OK, I think I understand your idea now.
From a fresh glimpse at the implementation we might indeed replace the `vector_constructor` which a new separate `collection_constructor::style_type`, which would be used only after preparation (as `vector_constructor` is used now).

I'm going to look into it deeper from now on, but high hopes it can be done the way with additional style type.
Feels like a good solution for me as well, since I noticed it. 🤞

QuerthDP replied:
I've just rebased the mentioned commits to implement vector style type instead of vector constructor.

Nevertheless I have two new ideas how could we improve the code from now on:
1. We might introduce the `list` style type, which would be used exactly the same as `vector` style type now (i.e. only after preparation). It would leave us with a straight look at what is **truly** a list, and what could be also a vector.
2. We could revert the rename of `list` to `list_or_vector` and provide more comments about the vector using `list` style before the preparation (feels a bit misleading, but would be also a good idea, and it requires less changes to the code).

Please take a look at these changes and let me know what do you think!

QuerthDP replied:
@piodul please take a look at the changes as well, as we discussed how the implementation of the expressions should look like.

avikivity commented:
Looks generally good. I did not like the addition of vector_constructor, it looks like another special case of collection_constructor (even if the name isn't an exact match) and doesn't merit forcing everyone who implements a visitor to special-case it.

I'd like to see tests that check that an INSERT with a bind variable and a vector with incorrect length fails validation.

nyh commented on test/cqlpy/cassandra_tests/validation/operations/cql_vector_test.py:
```diff
@@ -0,0 +1,478 @@
+#         assertRows(execute("SELECT f([1.0, 2.0], value) FROM %s"), expected);
+#     }
+
+# Test commented out because Scylla doesn't support token(*literal*) syntax.
```
This is https://github.com/scylladb/scylladb/issues/12607 or https://github.com/scylladb/scylladb/issues/5411. I think this test can be uncommented and xfail with the reason listed as `#5411` or something.

nyh commented:
Hi, we have an issue https://github.com/scylladb/scylladb/issues/19455 about adding Cassandra 5's "VECTOR" type to Scylla.
If this PR implements this feature as this issue requests (namely, Cassandra 5 compatibility) please mention "Fixes #19455" in the cover letter. If it only partially implements this feature, please explicitly mention what still needs to be done or what's different, and mention "Refs #19455" (which connects the patch to the issue, but won't close the issue automatically).

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22488

QuerthDP commented:
Changelog:

- adjusted documentation for vector support for Lua 
- removed unnecessary support for `empty` type in vectors
- added description of vector type in proper commit message

janpiotrlakomy commented on test/cqlpy/cassandra_tests/validation/operations/cql_vector_test.py:
```diff
@@ -0,0 +1,478 @@
+from ...porting import *
+
+
+# Some of the lines in these tests are commented out because Scylla doesn't support arithmetic operations in literals.
```
Yes, your understanding is correct and thanks for the context!

The only other meaningful changes in the translated tests  (except for 1 accidental one that I just found), are the error messages in the `assertInvalidThrowMessage` assertions. That is because some of the error messages differ between Scylla and Cassandra, so we have to test a subset of each message, e.g.  check for "bytes" instead of "Unexpected 2 extraneous bytes after vector<text, 2> value".

As for the `1+1` syntax, I think we will add a new cqlpy test that covers it and mark it xfail on 2693.

janpiotrlakomy commented on test/cqlpy/cassandra_tests/validation/operations/cql_vector_test.py:
```diff
@@ -0,0 +1,478 @@
+                                    "INSERT INTO %s (pk, value) VALUES (0, ?)", [1, 2, 3])
+
+
+@pytest.mark.skip(reason="The Python driver doesn't support variable width types in vectors yet")
```
This feature is supported by the datastax python driver (commit https://github.com/datastax/python-driver/commit/c4a808d58aa3d6b64bd491f24548b47c716b0185), but hasn't been added yet to the Scylla python driver.

As of right now, Scylla python driver only supports 4 byte types inside vectors (commit https://github.com/scylladb/python-driver/commit/6b46906927a8518a0a4f02625c9f49b99321248b), and I don't think there is an issue open asking to add support for arbitrary subtypes.

Regarding the other drivers, the Cassandra Java driver just added support for this (commit https://github.com/apache/cassandra-java-driver/commit/2e0c44c020819c928730e1aac812bf2f475d8256).

nyh replied:
What I did in the past in such a case is what you can see in driver_bug_1() in conftest.py - skip the test on a driver version which doesn't support it, but **run** it at least once, on your developer machine, on a driver version that does support it. I know this means your test won't be a good regression test because at least for now it will be skipped on the CI machines (which run Scylla's driver), but it will allow you to be sure that you actually tested this feature. With this part of the test completely commented out and not tested even once, you don't know if you implemented this part of the code correctly - and if you run it once with Cassandra's driver, you will know.

And please open an issue on our Python driver saying there is a missing feature that needs to be copied from Cassandra's driver. Thanks.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22488

QuerthDP commented:
Changelog:
- removed `vector_constructor` and implemented vector style type instead

nyh commented on concrete_types.hh:
```diff
@@ -15,6 +15,7 @@
 #include "types/map.hh"
 #include "types/set.hh"
```
What do you actually mean by "fixed" length? I am guessing (please correct me if I'm wrong) that all you meant by "fixed" is that there is no "append" operation (unlike list<>, whose bread-and-butter is this append operation). However, it's not "fixed" in that can assign to a vector column a vector or *any* length and even assign a new vector with a different length to a column that previously held a vector of some other length. Right?

I hope this is explained in more detail somewhere (I'm only starting to read this series now, so forgive me if it's already explained better elsewhere).

QuerthDP replied:
Yes that's right. The "fixed" length here only informs about the fixed number of dimensions (elements) of a vector.
The actual size may vary depending on what type of elements is stored inside.

QuerthDP replied:
> and even assign a new vector with a different length to a column that previously held a vector of some other length. Right?

Although I would like to deny this. The vector column have to be declared with a concrete length, so that there is not possible to assign a vector of f.e. length 3 to a column of vectors of length 5.

nyh replied:
To my above comment I wanted to add another question - are vectors values mutable or immutable? In other words, can I do something like `set v[3] = 2`? I am guessing (maybe I'll see this answered later) that the answer is that no, I can't do that, because it's not clear how to handle `set v[3] = 2` if the vector value doesn't already have 3 elements (and you said that "fixed length" thing).

nyh replied:
> Yes that's right. The "fixed" length here only informs about the fixed number of dimensions (elements) of a vector. The actual size may vary depending on what type of elements is stored inside.

Oh, wow. I didn't expect that. It is a "vector" in a mathematical vector-space sense, not a vector in the C++ or Java sense. I tried to imagine some lenient meaning of "fixed length" but in fact it is fixed in a very very strict sense - if you have a `VECTOR <FLOAT, 5>` it can only ever include 5-dimensional vectors (i.e., list of 5 floats), and nothing else.

So your comment about "fixed length" is 100% correct, just an example like `VECTOR <FLOAT, 5>` might have made it clearer just how "fixed" this "fixed" really is.

QuerthDP replied:
The set operation is not supported in vectors, as you have guessed correctly.

QuerthDP replied:
> just an example like VECTOR <FLOAT, 5> might have made it clearer

I believe it is explained clearer in the commit message introducing CQL syntax support, but I can surely agree it would have been a helpful tip if we have mentioned that before in this conversation.

nyh commented on test/boost/user_function_test.cc:
```diff
@@ -769,12 +795,15 @@ SEASTAR_TEST_CASE(test_user_function_nested_types) {
         e.execute_cql("INSERT INTO my_table (key, val) VALUES ('foo', 3);").get();
-        e.execute_cql("CREATE FUNCTION my_func(val int) CALLED ON NULL INPUT RETURNS map<int, frozen<set<frozen<list<frozen<tuple<text, bigint>>>>>>> LANGUAGE Lua AS  \
-                      'return {[42] = {[{{\"foo\", 41}, {\"bar\", 40}}] = true, [{{\"bar\", 40}}] = true}, [39] = {}}';").get();
+        e.execute_cql("CREATE FUNCTION my_func(val int) CALLED ON NULL INPUT RETURNS map<int, frozen<set<frozen<list<frozen<tuple<text, frozen<vector<bigint, 2>>>>>>>>> LANGUAGE Lua AS  \
```
Why "frozen<vector<>>" here with that "frozen"?
Is this just to test that frozen<vector> is allowed? If the answer is yes, I think you should have a separate test for it (in cqlpy, please, so you can also check Cassandra) and not in the middle of some UDF test. If the answer is no - i.e., frozen is mandatory here, I wonder why is in mandatory... Should it be mandatory? Is it also mandatory in Cassandra?

QuerthDP replied:
Good catch from you! 👀
I believe the `frozen<...>` tag was left there by mistake.
As we began the implementation we wrapped vector in frozen in every test we introduced, which was later dismissed.

Neither it was the purpose to test the `frozen<...>` tag here, nor it is mandatory.

nyh commented on test/boost/user_types_test.cc:
```diff
@@ -628,6 +629,52 @@ SEASTAR_TEST_CASE(test_nonfrozen_user_types_prepared) {
     });
 }
 
+// This test is a boost version of sandwichBetweenUDTs test from Apache Cassandra.
```
Since I believe you also translated the same test to cqlpy, I think this version is redundant, and it's worse (among other things, because you can't run it against Cassandra) so I think it's better to just remove this one.

nyh commented on transport/server.cc:
```diff
@@ -1952,6 +1953,11 @@ class type_codec {
         }
+        if (type->is_vector()) {
+            r.write_short(uint16_t(type_id::CUSTOM));
+            r.write_string(type->name());
```
I'm surprised that they didn't add a new type id. Maybe it's a good precedent for us: In the future, we can add more types that we can invent as a Scylla extension and encode them in the transport in exactly this way, and we won't need to worry about breaking Cassandra compatibility (since we can't just pick type 0x22, in fear one day Cassandra will start to use it).

avikivity replied:
It's not surprising, the type was first added to Datastax' fork (which couldn't claim a new type ID), then was contributed to Cassandra without changing the type ID because it would break drivers.

nyh replied:
But as you can see, this never happened before. I'm pretty sure the Datastax guys could have coordinated with Apache guys (same guys? :-)) to pick a number. Anyway, as I said, it's a good precedent in case in the future we'll want to add our own types. Too bad other things other aspects of the protocol don't have this CUSTOM trick - for example once I had the idea to add a new consistency level, but there is no way to do it without picking a number and hoping Cassandra don't pick the same number in the future.

nyh commented on docs/cql/types.rst:
```diff
@@ -641,6 +643,37 @@ Unlike other "composed" types (collections and UDT), a tuple is always ``frozen<
+it is not possible to update only some elements of a vector (without updating the whole vector).
+
+Nevertheless, types stored in a vector are not implicitly frozen, so if you want to store a frozen collection or 
+frozen UDT in a vector, you need to explicitly wrap them using `frozen` keyword.
```
The last paragraph is interesting. I think the word "neverthless" is a bit confusing (it confused me for a second) because this paragraph doesn't actually contradict to the previous paragraph and in my opinion - not even related to it. The previous paragraph is about `frozen<vector<...>`, and this one is about `vector<frozen<...>`. In any case, I wonder if anyone would ever ever use a vector of anything except float or double :-)

QuerthDP replied:
Indeed it might be a bit confusing at a first glance.
As you mentioned the core purpose for vectors are for storing floats and doubles, so I think this paragraph is unnecessary and may be removed.  Don't you agree?

QuerthDP replied:
AFAIK Cassandra did exactly that what we implemented, i.e. it allows both frozen and unfrozen collections without implicitly freezing them.

In this case, as you suggested, we might only change the misleading vocabulary and it should be just fine.

avikivity replied:
The nested collections are implicitly frozen, even if the "frozen" flag isn't set.

Frozen means individual cells cannot be updated independently (`UPDATE tab SET my_set = my_set + {3}`).

There's no syntax  (or support in the storage engine) to update a nested set element, so it's implicitly frozen.

Similarly, a vector is itself implicitly frozen (you cannot update a single element), so any nested collections would be similarly implicitly frozen.

I guess we have to distinguish between "visible frozen" (shows up with frozen\<> wrappers) and implicitly frozen (the entire collection is atomic). Yuch.

QuerthDP replied:
That's all true. I might have used the wrong vocabulary here as well.

What I really meant by `implicitly freezing` was a recursive call of `freeze()` function on the element type (like it is done in both frozen collections and tuples).

The lack of the recursive call implies that the types inside the vector are not visibly frozen, while efectively being frozen (cannot be updated) as you stated.

nyh reviewed (commented):
The whole series looks very good to me, I only left a few minor comments questions and requests.

QuerthDP commented on cql3/Cql.g:
```diff
@@ -1951,6 +1951,7 @@ comparator_type [bool internal] returns [shared_ptr<cql3_type::raw> t]
     : n=native_or_internal_type[internal]     { $t = cql3_type::raw::from(n); }
     | c=collection_type[internal]   { $t = c; }
     | tt=tuple_type[internal]       { $t = tt; }
+    | vt=vector_type                { $t = vt; }
```
Thank you for profound explanation! Now I'm sure it was correct to remove that! 🙇

QuerthDP commented on cql3/cql3_type.cc:
```diff
@@ -305,6 +305,56 @@ class cql3_type::raw_tuple : public raw {
+    size_t _dimension;
+
+    // This limitation is acquired from the maximum number of dimensions in OpenSearch. 
+    static constexpr size_t MAX_VECTOR_DIMENSION = 16000;
```
In fact, as we checked in Cassandra's code, the limit was never respected. It is just a note in documentation, which we believe is undervalued, and we could support vectors of higher lengths.

We have used the limit from OpenSearch as reference as we'd like to introduce OpenSearch to Scylla's toolchain to support ANN search (which is the main reason we add vector type).

nyh replied:
Funny about the incorrect documentation :-)

By the way, why do we even need this limitation (which you said Cassandra doesn't actually have)? Is it just  a "guardrail" - because if someone has million-dimensional vectors they will have lousy performance and this is likely a typo and not intended?
We could make this limitation a configuration option, then we don't even need to document it (the failed CREATE TABLE will tail you what configuration option you need to change) and it won't matter if our default matches OpenSearch or not.

In any case, please check like I suggested in another comment what happens for negative integers - we definitely don't want -1 to be treated as 2147483647 and then if we don't have a maximum, this number is respected :-)

QuerthDP replied:
> Is it just a "guardrail" - because if someone has million-dimensional vectors they will have lousy performance and this is likely a typo and not intended?

That's correct, that was the idea behind adding the limitation.

> We could make this limitation a configuration option, then we don't even need to document it (the failed CREATE TABLE will tail you what configuration option you need to change) and it won't matter if our default matches OpenSearch or not.

Thanks for the advice, I'll take a look at it, but I believe it's not that high priority now.

Anyway it might be better not to mention OpenSearch limits here anyway, don't you think?

> In any case, please check like I suggested in another comment what happens for negative integers - we definitely don't want -1 to be treated as 2147483647 and then if we don't have a maximum, this number is respected :-)

It of course needs to be checked, and hopefully this will be resolved soon.

nyh replied:
That's why I called this a "guardrail", not a "configuration option" - guardrails are more fashionable :-)

janpiotrlakomy commented on cql3/expr/expression.cc:
```diff
@@ -1382,6 +1383,20 @@ static managed_bytes reserialize_value(View value_bytes,
+                elements[i] = reserialize_value(managed_bytes_view(elements[i]), elements_type);
+            }
+        }
+
```
If my understanding is correct Cassandra supports vectors of any data type ([source](https://cassandra.apache.org/doc/latest/cassandra/reference/vector-data-type.html)), but I don't think I understand what you mean by scalar types , could you elaborate on that?

nyh replied:
I think Avi thought that we shouldn't support `vector<list<int>>,3>` and such. But IIUC Cassandra does support this bizarre use case, so perhaps we should do (although I don't know why).
We definitely should get rid of protocol version 2 support (this is https://github.com/scylladb/scylladb/issues/9791), but I don't think that it should be done in the middle of this series.

avikivity replied:
Yes. Support is very poor though:

```
[cqlsh 6.3.0 | Cassandra 5.1-SNAPSHOT | CQL spec 3.4.8 | Native protocol v5]
Use HELP for help.
cqlsh> CREATE TABLE ks.v2(id int primary key, vl vector<list<int>, 3>, vm vector<map<int, int>, 4>);
cqlsh> SELECT * FROM ks.v2;

 id | vl | vm
----+----+----

(0 rows)
cqlsh> INSERT INTO ks.v2(id, vl, vm) VALUES (1, [[1], [], [4, 7]], [{}, {1: 2}, {3: 4, 5: 8}, {}]);
cqlsh> SELECT * FROM ks.v2;
Traceback (most recent call last):
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 761, in recv_results_rows
    self.parsed_rows = [decode_row(row) for row in rows]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 761, in <listcomp>
    self.parsed_rows = [decode_row(row) for row in rows]
                        ^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 758, in decode_row
    return tuple(decode_val(val, col_md, col_desc) for val, col_md, col_desc in zip(row, column_metadata, col_descs))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 758, in <genexpr>
    return tuple(decode_val(val, col_md, col_desc) for val, col_md, col_desc in zip(row, column_metadata, col_descs))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 755, in decode_val
    return col_type.from_binary(raw_bytes, protocol_version)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/cqltypes.py", line 305, in from_binary
    return cls.deserialize(byts, protocol_version)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/cqltypes.py", line 1424, in deserialize
    raise VectorDeserializationFailure("Cannot determine serialized size for vector with subtype %s" % cls.subtype.__name__)
cassandra.VectorDeserializationFailure: Cannot determine serialized size for vector with subtype ListType(Int32Type)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 766, in recv_results_rows
    decode_val(val, col_md, col_desc)
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 755, in decode_val
    return col_type.from_binary(raw_bytes, protocol_version)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/cqltypes.py", line 305, in from_binary
    return cls.deserialize(byts, protocol_version)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/cqltypes.py", line 1424, in deserialize
    raise VectorDeserializationFailure("Cannot determine serialized size for vector with subtype %s" % cls.subtype.__name__)
cassandra.VectorDeserializationFailure: Cannot determine serialized size for vector with subtype ListType(Int32Type)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/avi/cassandra/bin/../pylib/cqlshlib/cqlshmain.py", line 892, in perform_simple_statement
    result = future.result()
             ^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/cluster.py", line 4956, in result
    raise self._final_exception
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/connection.py", line 1263, in process_msg
    response = decoder(header.version, self.user_type_map, stream_id,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 1202, in decode_message
    msg = msg_class.recv_body(body, protocol_version, user_type_map, result_metadata, cls.column_encryption_policy)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 739, in recv_body
    msg.recv(f, protocol_version, user_type_map, result_metadata, column_encryption_policy)
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 725, in recv
    self.recv_results_rows(f, protocol_version, user_type_map, result_metadata, column_encryption_policy)
  File "/home/avi/cassandra/bin/../lib/cassandra-driver-internal-only-3.29.0.zip/cassandra-driver-3.29.0/cassandra/protocol.py", line 768, in recv_results_rows
    raise DriverException('Failed decoding result column "%s" of type %s: %s' % (col_md[2],
cassandra.DriverException: Failed decoding result column "vl" of type org.apache.cassandra.db.marshal.VectorType<list, 3>: Cannot determine serialized size for vector with subtype ListType(Int32Type)

```

Vectors of VARINT also fail.

I think we should reject any vector of types that have non-fixed size for now (can use value_length_if_fixed()).

QuerthDP replied:
I think that is because you're using cassandra-driver 3.29.0.
AFAIK the support for variable length types has been introduced in 3.29.2
@janpiotrlakomy please confirm, as you know the details better.

avikivity replied:
If it does work in a newer driver, great. Please work with https://github.com/scylladb/python-driver team to update the driver to support the new encoding.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22488

janpiotrlakomy commented:
Changelog:
- changed cql_vector_test so that features not supported by the Scylla driver can be tested using the Datastax driver
- added relevant issue numbers to comments and marked xfail a previously commented out test in cql_vector_test

QuerthDP commented on test/boost/cql_query_test.cc:
```diff
@@ -1664,6 +1699,99 @@ SEASTAR_TEST_CASE(test_tuples) {
+            return *schema_builder(ks_name, "cf")
+                    .with_column("id", int32_type, column_kind::partition_key)
+                    .with_column("v", vt)
+                    .build();
```
The reason is quite simple, we wanted to mimic the test_tuples above 😅
Not much critical thinking was involved in that 😶‍🌫️

QuerthDP commented on cql3/cql3_type.cc:
```diff
@@ -305,6 +305,56 @@ class cql3_type::raw_tuple : public raw {
+    }
+
+    virtual void freeze() override {
+        _frozen = true;
```
https://github.com/scylladb/scylladb/pull/22488#discussion_r1932625269

I hope the response resolves your doubts.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=22488

QuerthDP commented:
Changelog:
- adjust docs (used @nyh's description of a vector. Thank you!)
- disallow negative numbers as vector dimension
- remove unexpected `freeze<>` in one of JSON tests

QuerthDP commented on docs/cql/types.rst:
```diff
@@ -641,6 +643,37 @@ Unlike other "composed" types (collections and UDT), a tuple is always ``frozen<
+Vectors
+^^^^^^^
+
+A ``vector`` is an array of elements of the same type with a fixed size. 
```
I've used your definition, which seems really descriptive. I hope you don't mind 👉👈

QuerthDP commented:
I've noticed that the CI didn't pass on a single test.

>Failed Tests (1/43308):
[cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14841/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_secondary_index_test_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/secondary_index_test.debug.1)

As I don't have access to read the jenkins logs, could I ask anyone responsible to let me know what was the reason for the failure?
Thank you in advance.

nyh commented:
> As I don't have access to read the jenkins logs, could I ask anyone responsible to let me know what was the reason for the failure? Thank you in advance.

This is cqlpy, `cassandra_tests/validation/entities/secondary_index_test.py::testCreateAndDropIndexWithCamelCaseIdentifier` the test crashed Scylla once in debug build, on an AddressSanitizer failure. Almost surely unrelated to your patch.

nyh commented:
> > As I don't have access to read the jenkins logs, could I ask anyone responsible to let me know what was the reason for the failure? Thank you in advance.
> 
> This is cqlpy, `cassandra_tests/validation/entities/secondary_index_test.py::testCreateAndDropIndexWithCamelCaseIdentifier` the test crashed Scylla once in debug build, on an AddressSanitizer failure. Almost surely unrelated to your patch.

It appears to be the same as https://github.com/scylladb/scylladb/issues/22040. I'll update that issue with more information.

Jadw1 commented:
Retriggered CI

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14876/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14876/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_functions_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/expr_test
:small_blue_diamond: boost/json_cql_query_test
:small_blue_diamond: boost/types_test
:small_blue_diamond: boost/user_function_test
:small_blue_diamond: boost/user_types_test
:small_blue_diamond: boost/view_schema_test
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6224/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3104/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14236/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14876/)
#### Failed Tests (36/43900):
- [test_no_read_amplification_on_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/read_amplification_test/TestReadAmplification/Tests___dtest_with_topology_changes___test_no_read_amplification_on_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_read_amplification_on_repair)
- [test_cluster_reduction_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/cdc_test/TestCdc/Tests___dtest_with_topology_changes___test_cluster_reduction_with_cdc_Multi_DC_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_reduction_with_cdc[Multi_DC_cluster])
- [test_multi_dc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/lwt_test/TestLwt/Tests___dtest_with_topology_changes___test_multi_dc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multi_dc)
- [test_cluster_reduction_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/cdc_test/TestCdc/Tests___dtest_with_topology_changes___test_cluster_reduction_with_cdc_Multi_DC_cluster__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_reduction_with_cdc[Multi_DC_cluster])
- [test_repair_option_pr_dc_host](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_option_pr_dc_host) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_option_pr_dc_host)
- [test_safe_unsafe_rebuild[2]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest_with_topology_changes___test_safe_unsafe_rebuild_2_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_safe_unsafe_rebuild[2])
- [test_safe_unsafe_rebuild[2]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest_with_topology_changes___test_safe_unsafe_rebuild_2__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_safe_unsafe_rebuild[2])
- [test_data_distribution_balance[6-IncrementalCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_topology_changes___test_data_distribution_balance_6_IncrementalCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[6-IncrementalCompactionStrategy])
- [test_data_distribution_balance[6-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_topology_changes___test_data_distribution_balance_6_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[6-TimeWindowCompactionStrategy])
- [test_lwt_truncate](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14876/testReport/junit/lwt_schema_modification_test/TestLWTSchemaModification/Tests___dtest_with_topology_changes___test_lwt_truncate) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lwt_truncate)
#### Build Details:
- Duration: 11 hr
- Builder: i-05bc3aef4e87c2b58 (m5ad.8xlarge)

Jadw1 commented:
Failed tests look unrelated, some of them are known (like https://github.com/scylladb/scylladb/issues/20448), retriggered.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15029/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15029/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_functions_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/expr_test
:small_blue_diamond: boost/json_cql_query_test
:small_blue_diamond: boost/types_test
:small_blue_diamond: boost/user_function_test
:small_blue_diamond: boost/user_types_test
:small_blue_diamond: boost/view_schema_test
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6351/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3234/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14366/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15029/)
#### Build Details:
- Duration: 10 hr
- Builder: i-0d778f093c96f4304 (m5ad.8xlarge)

avikivity commented on types/types.cc:
```diff
@@ -939,6 +956,7 @@ is_fixed_size_int_type(const abstract_type& t) {
     case k::utf8:
     case k::uuid:
     case k::varint:
+    case k::vector:
```
A default returning false means that when we add a new kind it will return false.

Without a default, one sees a build error and has to fix it, prompting them to make a decision.

QuerthDP replied:
>Without a default, one sees a build error and has to fix it, prompting them to make a decision.

That's exactly how we managed to notice it!

avikivity commented on types/types.cc:
```diff
@@ -1610,6 +1632,124 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+        : concrete_type(kind::vector, make_name(elements, dimension),
+        elements->value_length_if_fixed() ? std::optional(elements->value_length_if_fixed().value()*dimension):std::nullopt),
+        _elements_type(elements), _dimension(dimension) {
+    _contains_set_or_map = _elements_type->contains_set_or_map();
```
I wanted to ban vectors of collections too, but it turns our Cassandra allows them.

avikivity commented:
Queued
